src/unitxt/operators.py

```diff
@@ -213,6 +213,103 @@ def process(
         return flatten_dict(instance, parent_key=self.parent_key, sep=self.sep)
 
 
+class ModelInputFormatter(StreamInstanceOperator):
+    r"""Generates the whole input to the model, from constant strings that are given as args, and from values found in specified fields of the instance.
+
+    WholeInputFormatter expects its input instance to have been processed by a Template StreamInstanceOperator
+    (from templates.py), or any extension thereof, and hence to contain a field named "source", and to have
+    been processed by RenderInstruction StreamInstanceOperator (from renderer.py) and hence to contain a non None
+    field "instruction".
+    Demos are assumed to be included as well, in a field whose name is specified by argument 'demos_field'. That
+    field is assumed to contain a list of dicts, each containing keys: "source" and "target" of a single demo.
+    And finally, system_prompt is also assumed to be included
+    in the instance, in a field whose name is specified by argument 'system_prompt'.
+    The value in field "source" of the input instance describes (verbalizes) original values in the instance (as read
+    from the source dataset), in the context of the underlying task.
+    WholeInputFormatter makes this (verbalization of the) input into a whole-input-to-the-model, by combining it
+    with instruction, demos, and system_prompt, read from the input instance, through both formatting args
+    formatting arguments.
+    Following the Renderers action, the formatted model input is fed into field "source" (overwriting the field)
+    Also following renderers, this operator pops fields "instruction" and "demos" from the input instance.
+
+    Args:
+        demos_field (str): the name of the field that contains the demos, being dicts with "source" and "target" keys
+        system_prompt (str): the name of the field containing an introductory text and or xml tags
+        demo_format (str): formatting string for a single demo, combining fields source and target
+        model_input_format (str) overall output format, combining system_prompt, instruction, demos, and source
+    """
+
+    demos_field: str = "demos"
+    system_prompt: str = None
+    demo_format: str = (
+        "{source}\n{target}\n\n"  #  example: "User: {source}\nAgent: {target}\n\n"
+    )
+    model_input_format: str = (
```

Comment:
Why are the `system_prompt` and `instruction` fields of the instance and not explicit parts of this parameter?

And then the formatter could be used by setting `model_input_format` to (for example):

"<SYS>whaterver the system prompt is</SYS>\n whatever additional instructions there are \n{demos}\n{source}\n"

In other words, i would opt for minimizing the things that are passed by the instance to a bare minimum.

Reply:
Yes. Starting with removing `system_prompt` altogether.
The real things: `source` and `instruction`, and `template` (the last two separated or joint), and the `demos`, are expected to live in the instance from their birth to the creation of this `model input`, at which time they are popped from the instance. Part of the evolution of the stream.

Reply:
@dafnapension @yoavkatz I would separate between things that depend on the data and things that don't, hence IMO having `demos` on the instance makes sense.

However for the `instruction`, i am not sure. Since its a fixed value for all instances, i think it should be passed directly to the formatter (or mega operator) as a parameter. As for `template`, haven't seen that in your code, i think? Is it another field set on the instance?

Reply:
Hi @matanor , `template` is a stream operator, assumed to be applied before our  `ModelInputFormatter`, and, among others, prepare the demo records, with their expected fields `source` and `target`

Reply:
oh ok, thanks! i was confused, thought its another field on the instance ...